duolingo.py

```diff
@@ -28,20 +28,84 @@ class DuolingoException(Exception):
 class Duolingo(object):
     USER_AGENT = "Python Duolingo API/{}".format(__version__)
 
-    def __init__(self, username, password):
-        self.username = username
-        self.password = password
-        self.user_url = "https://duolingo.com/users/%s" % self.username
+    _url_base = "https://{host}{base_path}"
+    _url_user_no_auth = "/users?username={username}"
+    _url_login = "/login?fields="
+
+    def __init__(self, username="", password="", host="www.duolingo.com", base_path="/2017-06-30"):
+        self.host = host
+        self.base_path = base_path
+        self.base_url = self._url_base.format(host=self.host, base_path=self.base_path)
+
         self.session = requests.Session()
         self.leader_data = None
-        self.jwt = None
+        self.jwt = None        
+
+        if username:
+            self.username = username
+            if password:
+                self.password = password
+                self._login()
+            self._get_user_data(username)
+        self.voice_url_dict = None
 
-        if password:
-            self._login()
+    def _get_user_data(self, username):
+        """Gets user data with _get_data if logged in
+            and _get_user_data_no_auth otherwise
+        
+        Arguments:
+            username {str} -- username on Duolingo
+        """        
+        self.username = username
+        self.user_url = "https://duolingo.com/users/%s" % self.username # TODO: should be removed
+
+        data = self._get_data() if self.jwt else self._get_user_data_no_auth(username)
```

Comment:
Just to check, are these endpoints returning data in the same format? I should probably check myself..

Reply:
Endpoints return text/json, methods return type `dict`

Reply:
Yeah, but the keys in those dictionaries seem to be completely different in my tests?

```
from duolingo import Duolingo
username = ...
password = ...
d1 = Duolingo(username)
k1 = set(list(d1.user_data.__dict__.keys()))
d2 = Duolingo(username, password)
d2._get_user_data(username)
k2 = set(list(d2.user_data.__dict__.keys()))
print(len(k1))  # 27
print(len(k2))  # 99
print(k1.intersection(k2))  # {'username', 'bio', 'id'}
```

You can see that with password, you get a lot more user data than without, but either way, they share very few keys, they're almost totally different sets of data

Reply:
It's different endpoint, I've just checked. 

I compared `{{ baseURL }}/users/{{ userID }}?fields=` with `{{ baseURL  }}/users?username={{ username }}` for logged in and non-logged in users respectively. That's what I used for the OpenAPI. Seems you use mostly different sets of API methods.

Reply:
I didn't actually realised that there are at least three endpoints to get user data and they are all different. But those with base path `2017-06-30` at least have similar structure. I've never seen the URLs without `2017-06-30` used. Maybe they are different API gateways?

Reply:
Yeah, so we probably can't just swap out one call with the other